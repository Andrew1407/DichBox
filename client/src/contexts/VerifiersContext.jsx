import React, { createContext, useState, useCallback, useReducer } from 'react';
import axios from 'axios';
import verifyDataReducer from '../reducers/verifyDataReducer';

export const VerifiersContext = createContext();

const VerifiersContextProvider = props => {
  const [warnings, setWarning] = useState({});
  const [correctInput, setCorrectState] = useState({});
  const [dataInput, dispatchDataInput] = useReducer(verifyDataReducer, {});
  const cleanWarningsClb = () => {
    setCorrectState({});
    setWarning({});
  };
  const cleanWarnings = useCallback(cleanWarningsClb, []);
  const setWarningsOnHandleClb = (wrngs, correctState) => {
    if (wrngs)
      setWarning({ ...warnings, ...wrngs });
    if (correctState)
      setCorrectState({ ...correctInput, ...correctState });
  };
  const setWarningsOnHandle = useCallback(setWarningsOnHandleClb, [warnings, correctInput]);
  const fetchBoxInput = async (username, boxName) => {
    const verifyBody = { username, boxName };
    const { data } = await axios.post(`${process.env.APP_ADDR}/boxes/verify`, verifyBody);
    return data;
  };
  const fetchUserInput = async (inputField, inputValue) => {
    const verifyBody = { inputField, inputValue };
    const { data } = await axios.post(`${process.env.APP_ADDR}/users/verify`, verifyBody);
    return data.foundValue;
  };
  const fetchPasswdVer = async (username, passwd) => {
    const passwdBody = { username, passwd };
    const { data } = await axios.post(`${process.env.APP_ADDR}/users/passwd_verify`, passwdBody);
    return data.checked;
  }
  const useVerifiersClb = verParams => {
    const verFields = Object.keys(verParams);
    const getVerifiersState = (arr = verFields) => arr.reduce(
      (isCorrectAll, isCorrectField) =>
        (isCorrectAll && !!correctInput[isCorrectField]), true
    );
    const getVerifier = field => {
      const verifierParams = verParams[field];
      if (!verifierParams) return;
      return async input => {
        const { fetchVerifier, defaultState } = verifierParams;
        let warningStyle = {
          borderColor: 'rgb(0, 255, 76)',
          text: ''
        };
        let isCorrect = true;
        const inputRegExp = verifierParams.regExp;
        if (defaultState?.value === input) {
          warningStyle = {
            borderColor: 'rgb(0, 217, 255)',
            text: ''
          };
          isCorrect = defaultState.correct;
        } else if (inputRegExp && !inputRegExp.test(input)) {
          warningStyle = {
            borderColor: 'crimson',
            text: verifierParams.warningRegExp
          };
          isCorrect = false;
        } else if (fetchVerifier) {
          const checkup = await fetchVerifier(input);
          if (checkup) {
            warningStyle = {
              borderColor: 'crimson',
              text: verifierParams.warningFetch
            }
            isCorrect = false;
          }
        }
        setWarningsOnHandle(
          { [field]: warningStyle },
          { [field]: isCorrect }
        );
      };
    };
    const updateUserDataClb = field => {
      const inputVerifier = getVerifier(field);
      return e => {
        e.preventDefault();
        const input = e.target.value;
        if (inputVerifier)
          inputVerifier(input);
        const dataReducerAction = {
          type: 'PUSH_DATA',
          data: { [field]: input }
        };
        dispatchDataInput(dataReducerAction);
      };
    };
    const getOnChangeVerifier = useCallback(updateUserDataClb, [dataInput, warnings, correctInput]);
    return { getVerifiersState, getOnChangeVerifier };
  };
  const useVerifiers = useCallback(useVerifiersClb, [warnings, correctInput, dataInput]);

  return (
    <VerifiersContext.Provider value={{ useVerifiers, fetchUserInput, warnings, correctInput, dataInput, dispatchDataInput, fetchPasswdVer, cleanWarnings, setWarningsOnHandle, fetchBoxInput }}>
      { props.children }
    </VerifiersContext.Provider>
  );
};

export default VerifiersContextProvider;
