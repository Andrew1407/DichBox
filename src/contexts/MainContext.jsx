import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios';
import { useHistory } from 'react-router-dom';

export const MainContext = createContext();

const MainContextProvider = props => {
  const history = useHistory();
  const getPathName = () => history.location.pathname
    .split('/')
    .slice(1, 2)[0];
  const [pathName, setPathName] = useState(getPathName());
  const [menuVisible, setMenuVisible] = useState(true);
  const [username, setUsername] = useState(null);
  const [userData, setUserData] = useState({});
  const [id, idState] = useState(localStorage.getItem('id'));
  const setId = id => {
    localStorage.setItem('id', id);
    idState(id);
  };

  useEffect(() => {
    const fetchUsername = async () => {
    if (id) {
      const { data } = await axios.post('http://192.168.0.223:7041/users/name', { id });
      setUsername(data.name);
    }
  };
    
  fetchUsername();
  }, [id]);

  useEffect(() => {
    const fetchUserData = async () => {
      if (pathName.length) {
        const findBody = { id, name: pathName };
        const { data } = await axios.post('http://192.168.0.223:7041/users/find', findBody);
        setUserData(data)
      }
    };
    
    fetchUserData();
  }, [pathName]);

  
  return (
    <MainContext.Provider value={{ menuVisible, setMenuVisible, id, setId, username, setUsername, userData, setPathName }}>
      {props.children}
    </MainContext.Provider>
  );
};

export default MainContextProvider;
