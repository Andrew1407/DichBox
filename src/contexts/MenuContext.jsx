import React, { createContext, useState, useReducer, useCallback } from 'react';
import filesReducer from '../reducers/filesReducer';

export const MenuContext = createContext();

const MenuContextProvider = props => {
  const [menuOption, setMenuOption] = useState('default');
  const [menuVisible, setMenuVisible] = useState(true);
  const [openedFiles, dispatchOpenedFiles] = useReducer(filesReducer, []);
  const [usersList, setUsersList] = useState(null);
  const [searchStr, setSearchStr] = useState(null);

  const searchFileInOpenedListClb = (searchName, searchPath) => {
    for (const index in openedFiles) {
      const { name, filePath, opened } = openedFiles[index];
      const found = searchName == name && searchPath === filePath;
      if (found)
        return { index, opened };
    }
    return { index: -1 };
  };
  const searchFileInOpenedList = useCallback(searchFileInOpenedListClb, [openedFiles]);

  return (
    <MenuContext.Provider value={{ searchStr, setSearchStr, usersList, setUsersList, searchFileInOpenedList, openedFiles, dispatchOpenedFiles, menuVisible, setMenuVisible, menuOption, setMenuOption }}>
      { props.children }
    </MenuContext.Provider>
  );
};

export default MenuContextProvider;
