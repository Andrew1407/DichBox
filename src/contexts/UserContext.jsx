import React, { createContext, useState, useEffect, useReducer } from 'react';
import axios from 'axios';
import { useHistory } from 'react-router-dom';
import  userDataReducer from '../reducers/userDataReducer';
import  usernameReducer from '../reducers/usernameReducer';

export const UserContext = createContext();

const UserContextProvider = props => {
  const history = useHistory();
  const getPathName = () => history
    .location
    .pathname
    .split('/')[1];
  const [pathName, setPathName] = useState(getPathName());
  const [username, dispatchUsername] = useReducer(usernameReducer, localStorage.getItem('name'));
  const [userData, dispatchUserData] = useReducer(userDataReducer, {});
  
  useEffect(() => {
    const fetchUserData = async () => {
      if (pathName && pathName.length) {
        const findBody = { username, pathName };
        const { data } = await axios.post('http://192.168.0.223:7041/users/find', findBody);
        dispatchUserData({ type: 'REFRESH_DATA', data });
      }
    };
    
    fetchUserData();
  }, [pathName]);

  return (
    <UserContext.Provider value={{ username, dispatchUsername, dispatchUserData, userData, pathName, setPathName }}>
      { props.children }
    </UserContext.Provider>
  );
};

export default UserContextProvider;
