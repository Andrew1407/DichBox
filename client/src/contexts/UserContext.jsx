import React, { createContext, useState, useEffect, useReducer, useContext } from 'react';
import axios from 'axios';
import { useHistory } from 'react-router-dom';
import { MenuContext } from './MenuContext';
import  userDataReducer from '../reducers/userDataReducer';
import  usernameReducer from '../reducers/usernameReducer';

export const UserContext = createContext();

const UserContextProvider = props => {
  const history = useHistory();
  const { setFoundErr, setLoading, setUsersList } = useContext(MenuContext);
  const getPathName = () => history
    .location
    .pathname
    .split('/')[1];
  const [pathName, setPathName] = useState(getPathName());
  const [username, dispatchUsername] = useReducer(usernameReducer, localStorage.getItem('name'));
  const [userData, dispatchUserData] = useReducer(userDataReducer, {});
  
  useEffect(() => {
    const fetchUserData = async () => {
      setUsersList(null);
      if (pathName && pathName.length) {
        setLoading(true);
        const findBody = { username, pathName };
        try {
          const { data } = await axios.post(`${process.env.APP_ADDR}/users/find`, findBody);
          dispatchUserData({ type: 'REFRESH_DATA', data });
        } catch (e) {
          if (!e.response) {
            const msg = 'It\'s a secret, but something terrible happened on the DichBox server...';
            setFoundErr(['server', msg]);
          } else {            
            const { status, data } = e.response;
            const errType = status === 404 ? 'user' : 'server'; 
            setFoundErr([errType, data.msg]);
          }
        } finally {
          setLoading(false);
        }
      }
    };
    
    fetchUserData();
  }, [pathName]);

  return (
    <UserContext.Provider value={{ username, dispatchUsername, dispatchUserData, userData, pathName, setPathName }}>
      { props.children }
    </UserContext.Provider>
  );
};

export default UserContextProvider;
