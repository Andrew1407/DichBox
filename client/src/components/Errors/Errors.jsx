import React, { useContext } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import { MenuContext } from '../../contexts/MenuContext';
import { componentMotion } from '../../styles/motions/menu-components';
import boxErrLogo from '../../styles/imgs/errors/box.png';
import dirErrLogo from '../../styles/imgs/errors/dir.png';
import serverErrLogo from '../../styles/imgs/errors/server.png';
import signInErrLogo from '../../styles/imgs/errors/signIn.png';
import userErrLogo from '../../styles/imgs/errors/user.png';
import signUpErrLogo from '../../styles/imgs/errors/signUp.png';
import '../../styles/errors.css';

const Errors = () => {
  const { foundErr } = useContext(MenuContext);
  const [logoKey, errMessage] = foundErr;
  const errLogos = {
    dir: dirErrLogo,
    box: boxErrLogo,
    user: userErrLogo,
    signIn: signInErrLogo,
    signUp: signUpErrLogo,
    server: serverErrLogo
  };

  return (
    <AnimatePresence>
    <motion.div
      { ...componentMotion }
      data-testid="errors-test" id="errors-container" className="menu-form"
    >
      <h1 id="error-title">Error</h1>
      <p id="error-message">{ errMessage }</p>
      <div id="error-image">
        <img src={ errLogos[logoKey] }/>
      </div>
    </motion.div>
    </AnimatePresence>
  );
}

export default Errors;
