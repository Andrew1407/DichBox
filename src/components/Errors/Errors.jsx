import React, { useContext } from 'react';
import { MenuContext } from '../../contexts/MenuContext';
import boxErrLogo from '../../styles/imgs/errors/box.png';
import dirErrLogo from '../../styles/imgs/errors/dir.png';
import serverErrLogo from '../../styles/imgs/errors/server.png';
import signInErrLogo from '../../styles/imgs/errors/signIn.png';
import userErrLogo from '../../styles/imgs/errors/user.png';
import signUpErrLogo from '../../styles/imgs/errors/signUp.png';
import '../../styles/errors.css';

const Errors = () => {
  const { foundErr } = useContext(MenuContext);
  const [logoKey, errMessage] = foundErr;
  const errLogos = {
    dir: dirErrLogo,
    box: boxErrLogo,
    user: userErrLogo,
    signIn: signInErrLogo,
    signUp: signUpErrLogo,
    server: serverErrLogo
  };

  return (
    <div className="menu-form">
      <h1 id="error-title"> Error</h1>
      <p id="error-message">{ errMessage }</p>
      <div id="error-image">
        <img src={ errLogos[logoKey] }/>
      </div>
    </div>
  );
}

export default Errors;
