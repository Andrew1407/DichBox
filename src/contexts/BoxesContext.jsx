import React, { createContext, useState, useContext, useCallback } from 'react';
import { useHistory } from 'react-router-dom';
import axios from 'axios';
import { MenuContext } from './MenuContext';
import { UserContext } from './UserContext';

export const BoxesContext = createContext();

const BoxesContextProvider = props => {
  const history = useHistory();
  const { userData, username } = useContext(UserContext);
  const {
    dispatchOpenedFiles,
    setFoundErr,
    setLoading
  } = useContext(MenuContext);
  const [boxesList, setBoxesList] = useState([]);
  const [listOption, setListOption] = useState('all');
  const [boxInfoHidden, setBoxHiddenState] = useState(false);
  const [boxDetails, setBoxDetails] = useState({});
  const [editBox, setEditBoxState] = useState(false);
  const [pathEntries, setPathEntries] = useState([]);

  const fetchEntriesClb = async (boxPath, initial) => {
    setLoading(true);
    const filesBody = {
      boxPath,
      viewerName: username,
      follower: userData.follower,
      editor: userData.editor,
      initial
    };
    try {
      const { data } = await axios.post(`${process.env.APP_ADDR}/boxes/files/list`, filesBody);
      const { type, dir, file } = data.entries;
      setLoading(false);
      if (type === 'dir') {
        setPathEntries(dir.src.length ? dir.src : [null]);
        return;
      }
      if (type === 'file')
        if (initial) setBoxErr(true);
        else dispatchOpenedFiles({ type: 'FILE_APPEND', file });
    } catch (e) {
      if (!e.response) {
        const msg = 'It\'s a secret, but something terrible happened on the DichBox server...';
        setFoundErr(['server', msg]);
      } else { 
        const { status, data } = e.response;
        const errType = status === 404 ? (boxPath.length > 2 ? 'dir' : 'box') : 'server';
        setFoundErr([errType, data.msg]);
      }
      setLoading(false);
    }
  };
  const fetchEntries = useCallback(fetchEntriesClb, [userData, history.location]);
  
  return (
    <BoxesContext.Provider value={{ fetchEntries, pathEntries, setPathEntries, boxesList, setBoxesList, listOption, setListOption, boxInfoHidden, setBoxHiddenState, boxDetails, setBoxDetails, editBox, setEditBoxState }} >
      { props.children }
    </BoxesContext.Provider>
  );
};

export default BoxesContextProvider;
