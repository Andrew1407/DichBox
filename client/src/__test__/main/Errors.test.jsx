import React from 'react';
import { cleanup, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Errors from '../../components/Errors/Errors';
import { MenuContext } from '../../contexts/MenuContext';
import boxErrLogo from '../styles/imgs/errors/box.png';
import dirErrLogo from '../styles/imgs/errors/dir.png';
import serverErrLogo from '../styles/imgs/errors/server.png';
import signInErrLogo from '../styles/imgs/errors/signIn.png';
import userErrLogo from '../styles/imgs/errors/user.png';
import signUpErrLogo from '../styles/imgs/errors/signUp.png';

describe('Errors tests', () => {
  const testData = [
    ['dir', 'dir error', dirErrLogo],
    ['box', 'box error', boxErrLogo],
    ['user', 'user error', userErrLogo],
    ['signIn', 'singIn error', signInErrLogo],
    ['signUp', 'singUp error', signUpErrLogo],
    ['server', 'server error', serverErrLogo]
  ];

  const tests = {
    'has title': () => {
      const foundErr = testData[0].slice(0, -1);
      const foundErrors = render(
        <MenuContext.Provider value={{ foundErr }}>
          <Errors />
        </MenuContext.Provider>
      ).getByTestId('errors-test');
      expect(foundErrors).toHaveTextContent('Error');
    },
    'has error content': () => {
      for (const testArgs of testData) {
        const foundErr = testArgs.slice(0, -1);
        const foundErrors = render(
          <MenuContext.Provider value={{ foundErr }}>
            <Errors />
          </MenuContext.Provider>
        ).getByTestId('errors-test');
        
        const [ _, errMsg, errImg ] = testArgs;
        const imgContainer = foundErrors.querySelector('#error-image');
        const imgEl = imgContainer.querySelector('img');

        expect(foundErrors).toHaveTextContent(errMsg);
        expect(imgContainer).toContainElement(imgEl);
        expect(imgEl).toHaveAttribute('src', errImg);
        cleanup();
      }
    },
    'matches snapshot': () => {
      const foundErr = testData[testData.length - 1].slice(0, -1);
      const tree = renderer.create(
        <MenuContext.Provider value={{ foundErr }}>
          <Errors />
        </MenuContext.Provider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    }
  };

  for (const key in tests)
    it(key, tests[key]);
});
