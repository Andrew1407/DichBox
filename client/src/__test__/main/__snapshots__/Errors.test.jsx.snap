// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Errors tests matches snapshot 1`] = `
<div
  className="menu-form"
  data-testid="errors-test"
  id="errors-container"
  style={
    Object {
      "transform": "translateX(-800px) translateZ(0)",
    }
  }
>
  <h1
    id="error-title"
  >
    Error
  </h1>
  <p
    id="error-message"
  >
    server error
  </p>
  <div
    id="error-image"
  >
    <img
      src="test-file-stub"
    />
  </div>
</div>
`;
